solutions/gold/cf-510C.mdx

```diff
@@ -0,0 +1,110 @@
+---
+id: cf-510C
+source: CF
+title: Fox And Names
+author: Daniel Ge
+---
+
+## Solution
+The key idea here is to perform topological sort on the characters of the alphabet. 
+
+We find what must come before what by comparing consecutive strings. If string $name_i$ is lexicographically less than $name_{i+1}$, we compare only the first occurence of a different letter. In other words, we take the common prefix out and compare the first letters. For example, if we are given,
+```
+axcd
+axbe
+```
+We see that the common prefix is "ax", so the first pair of differing letters is 'c' and 'b'. Since "axcd" comes first, that means 'c' comes before 'b' in the desired alphabet.
+
+In the special case that one word is the prefix of the other such as the case with,
+```
+abc
+abcde
+```
+We just check whether the second (lexicographically greater) string is longer as is supposed to be the case according to the problem statement. If this is indeed not the case, then we can just output "Impossible".
+## Implementation
+
+**Time Complexity:** $\mathcal{O}(n\text{ |name}_i|)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+// Used for topological sort
+vector<int> indegree = vector<int>(26, 0);
+vector<vector<int>> conn = vector<vector<int>>(26, vector<int>());
+vector<int> sorted;
+
+void topological_sort() {
+	queue<int> q;
+	for (int i = 0; i < 26; i++) {
+		if (indegree[i] == 0) { q.push(i); }
+	}
+	while (!q.empty()) {
+		int pos = q.front();
+		q.pop();
+		sorted.push_back(pos);
+		for (auto c : conn[pos]) {
+			if ((--indegree[c]) == 0) { q.push(c); }
+		}
+	}
+}
+
+int main() {
+	int n;
+	// We make sure each pair of consecutive strings are in the right order
+	string cur, prev;  
+
+	/*
+	 * No previous string to compare the first string to, 
+	 * so we read it directly into prev
+	 */
+	cin >> n >> prev;  
+	for (int i = 1; i < n; i++) {
+		cin >> cur;
+
+		// Maximum length to compare
+		int leng = min(prev.length(), cur.length());  
+		// How long the prefix has been the same for
+		int samed = 0;
+
+		while (samed < leng) {
```

Comment:
why not just use a for loop lmao?

Reply:
Mostly because samed is used outside the loop.

I was already considering changing it to a while loop and the other reviewer preferred while loop so I changed it.

Reply:
yeah a while loop makes more sense given the variable is also used outside